From e66fc563d82260258df95f659d1c7c2cf1adedfa Mon Sep 17 00:00:00 2001
From: Dominic Hargreaves <dom@earth.li>
Date: Wed, 11 Sep 2013 23:10:23 +0100
Subject: untaint version, if needed, in Module::Metadata

This change is imported from Module::Metadata 1.000018, exposed by
changes in Module::Load::Conditional, to fix taint mode use of this
module.

Bug: https://rt.cpan.org/Public/Bug/Display.html?id=88576
Bug-Debian: http://bugs.debian.org/722210
Origin: http://perl5.git.perl.org/perl.git/commit/bff978faaae3c8c03edc7cf579be6660fdc89fb3
Patch-Name: fixes/module_metadata_taint_fix.diff
---
 cpan/Module-Metadata/lib/Module/Metadata.pm |    4 +++-
 cpan/Module-Metadata/t/encoding.t           |    1 +
 cpan/Module-Metadata/t/lib/DistGen.pm       |    4 ++++
 cpan/Module-Metadata/t/lib/MBTest.pm        |    1 +
 cpan/Module-Metadata/t/lib/Tie/CPHash.pm    |    1 +
 cpan/Module-Metadata/t/metadata.t           |    1 +
 cpan/Module-Metadata/t/taint.t              |   29 +++++++++++++++++++++++++++
 cpan/Module-Metadata/t/version.t            |    1 +
 8 files changed, 41 insertions(+), 1 deletion(-)
 create mode 100644 cpan/Module-Metadata/t/taint.t

diff --git a/cpan/Module-Metadata/lib/Module/Metadata.pm b/cpan/Module-Metadata/lib/Module/Metadata.pm
index 4e416a8..224568b 100644
--- a/cpan/Module-Metadata/lib/Module/Metadata.pm
+++ b/cpan/Module-Metadata/lib/Module/Metadata.pm
@@ -613,7 +613,7 @@ sub _evaluate_version_line {
   # compiletime/runtime issues with local()
   my $vsub;
   $pn++; # everybody gets their own package
-  my $eval = qq{BEGIN { q#  Hide from _packages_inside()
+  my $eval = qq{BEGIN { my \$dummy = q#  Hide from _packages_inside()
     #; package Module::Metadata::_version::p$pn;
     use version;
     no strict;
@@ -626,6 +626,8 @@ sub _evaluate_version_line {
       };
   }};
 
+  $eval = $1 if $eval =~ m{^(.+)}s;
+
   local $^W;
   # Try to get the $VERSION
   eval $eval;
diff --git a/cpan/Module-Metadata/t/encoding.t b/cpan/Module-Metadata/t/encoding.t
index a0970e0..b010f7e 100644
--- a/cpan/Module-Metadata/t/encoding.t
+++ b/cpan/Module-Metadata/t/encoding.t
@@ -1,6 +1,7 @@
 #!perl
 
 use strict;
+use warnings;
 use File::Spec;
 use Test::More;
 
diff --git a/cpan/Module-Metadata/t/lib/DistGen.pm b/cpan/Module-Metadata/t/lib/DistGen.pm
index 9fbd6d0..2353120 100644
--- a/cpan/Module-Metadata/t/lib/DistGen.pm
+++ b/cpan/Module-Metadata/t/lib/DistGen.pm
@@ -1,6 +1,7 @@
 package DistGen;
 
 use strict;
+use warnings;
 
 use vars qw( $VERSION $VERBOSE @EXPORT_OK);
 
@@ -182,6 +183,7 @@ sub _gen_default_filedata {
       \$VERSION = '0.01';
 
       use strict;
+      use warnings;
 
       1;
 
@@ -205,6 +207,7 @@ sub _gen_default_filedata {
   $self->$add_unless('t/basic.t', undent(<<"    ---"));
     use Test::More tests => 1;
     use strict;
+    use warnings;
 
     use $self->{name};
     ok 1;
@@ -470,6 +473,7 @@ sub change_build_pl {
 
   $self->change_file( 'Build.PL', undent(<<"    ---") );
     use strict;
+    use warnings;
     use Module::Build;
     my \$b = Module::Build->new(
     # Some CPANPLUS::Dist::Build versions need to allow mismatches
diff --git a/cpan/Module-Metadata/t/lib/MBTest.pm b/cpan/Module-Metadata/t/lib/MBTest.pm
index 005920f..fb239ab 100644
--- a/cpan/Module-Metadata/t/lib/MBTest.pm
+++ b/cpan/Module-Metadata/t/lib/MBTest.pm
@@ -1,6 +1,7 @@
 package MBTest;
 
 use strict;
+use warnings;
 
 use IO::File ();
 use File::Spec;
diff --git a/cpan/Module-Metadata/t/lib/Tie/CPHash.pm b/cpan/Module-Metadata/t/lib/Tie/CPHash.pm
index b167622..217d642 100644
--- a/cpan/Module-Metadata/t/lib/Tie/CPHash.pm
+++ b/cpan/Module-Metadata/t/lib/Tie/CPHash.pm
@@ -20,6 +20,7 @@ package Tie::CPHash;
 
 require 5.000;
 use strict;
+use warnings;
 use vars qw(@ISA $VERSION);
 
 @ISA = qw();
diff --git a/cpan/Module-Metadata/t/metadata.t b/cpan/Module-Metadata/t/metadata.t
index b7adb1e..2503fad 100644
--- a/cpan/Module-Metadata/t/metadata.t
+++ b/cpan/Module-Metadata/t/metadata.t
@@ -3,6 +3,7 @@
 # vim:ts=8:sw=2:et:sta:sts=2
 
 use strict;
+use warnings;
 use lib 't/lib';
 use IO::File;
 use MBTest;
diff --git a/cpan/Module-Metadata/t/taint.t b/cpan/Module-Metadata/t/taint.t
new file mode 100644
index 0000000..ef527de
--- /dev/null
+++ b/cpan/Module-Metadata/t/taint.t
@@ -0,0 +1,29 @@
+#!/usr/bin/perl -T
+use strict;
+use warnings;
+
+use 5.008000;   # for ${^TAINT}
+use Test::More tests => 2;
+use Module::Metadata;
+use Carp 'croak';
+
+# stolen liberally from Class-Tiny/t/lib/TestUtils.pm - thanks xdg!
+sub exception(&) {
+    my $code = shift;
+    my $success = eval { $code->(); 1 };
+    my $err = $@;
+    return undef if $success;   # original returned ''
+    croak "Execution died, but the error was lost" unless $@;
+    return $@;
+}
+
+ok(${^TAINT}, 'taint flag is set');
+
+# without the fix, we get:
+# Insecure dependency in eval while running with -T switch at lib/Module/Metadata.pm line 668, <GEN0> line 15.
+is(
+    exception { Module::Metadata->new_from_module( "Module::Metadata" )->version },
+    undef,
+    'no exception',
+);
+
diff --git a/cpan/Module-Metadata/t/version.t b/cpan/Module-Metadata/t/version.t
index 061a063..e523f97 100644
--- a/cpan/Module-Metadata/t/version.t
+++ b/cpan/Module-Metadata/t/version.t
@@ -1,4 +1,5 @@
 use strict;
+use warnings;
 use Test::More;
 use Module::Metadata;
 use lib "t/lib/0_2";
